File: src/main/kotlin/school/charset/app/domain/exercise/ErrorType.kt
package school.charset.app.domain.exercise

// Stable identifiers for validation errors produced by `AnswerValidator`.
object ErrorType {
    object Answer {
        const val TYPE_MISMATCH = "answer.type-mismatch"
    }

    object Binary {
        const val EMPTY = "binary.empty"
        const val INVALID_CHARACTER = "binary.invalid-character"
        const val TOO_FEW_BITS = "binary.too-few-bits"
        const val TOO_MANY_BITS = "binary.too-many-bits"
        const val WRONG_VALUE = "binary.wrong-value"
    }

    object Format {
        const val EMPTY = "format.empty"
        const val UNKNOWN_CHOICE = "format.unknown-choice"
        const val WRONG_CHOICE = "format.wrong-choice"
    }

    object BitGroups {
        const val EMPTY = "bit-groups.empty"
        const val WRONG_GROUP_COUNT = "bit-groups.wrong-group-count"
        const val INVALID_CHARACTER = "bit-groups.invalid-character"
        const val WRONG_GROUP_LENGTH = "bit-groups.wrong-group-length"
        const val WRONG_VALUE = "bit-groups.wrong-value"
    }

    object HexBytes {
        const val EMPTY = "hex-bytes.empty"
        const val BYTE_OUT_OF_RANGE = "hex-bytes.byte-out-of-range"
        const val TOO_FEW_BYTES = "hex-bytes.too-few-bytes"
        const val TOO_MANY_BYTES = "hex-bytes.too-many-bytes"
        const val WRONG_VALUE = "hex-bytes.wrong-value"
    }

    object CodePoint {
        const val OUT_OF_RANGE = "code-point.out-of-range"
        const val SURROGATE = "code-point.surrogate"
        const val WRONG_VALUE = "code-point.wrong-value"
    }

    object Endianness {
        const val WRONG_CHOICE = "endianness.wrong-choice"
    }
}

File: lib/charset/exercise/error_type.ex
defmodule Charset.Exercise.ErrorType do
  @moduledoc """
  Stable identifiers for validation errors produced by `AnswerValidator`.

  These are business-event identifiers that the frontend consumes as i18n
  keys (`feedback.{error_type}.level{hint_level}`), NOT translations. Never
  use a string literal at a call site - always go through these functions, so
  a rename propagates everywhere (validator, tests, UI mapping).
  """

  def answer_type_mismatch, do: "answer.type-mismatch"

  def binary_empty, do: "binary.empty"
  def binary_invalid_character, do: "binary.invalid-character"
  def binary_too_few_bits, do: "binary.too-few-bits"
  def binary_too_many_bits, do: "binary.too-many-bits"
  def binary_wrong_value, do: "binary.wrong-value"

  def format_empty, do: "format.empty"
  def format_unknown_choice, do: "format.unknown-choice"
  def format_wrong_choice, do: "format.wrong-choice"

  def bit_groups_empty, do: "bit-groups.empty"
  def bit_groups_wrong_group_count, do: "bit-groups.wrong-group-count"
  def bit_groups_invalid_character, do: "bit-groups.invalid-character"
  def bit_groups_wrong_group_length, do: "bit-groups.wrong-group-length"
  def bit_groups_wrong_value, do: "bit-groups.wrong-value"

  def hex_bytes_empty, do: "hex-bytes.empty"
  def hex_bytes_byte_out_of_range, do: "hex-bytes.byte-out-of-range"
  def hex_bytes_too_few_bytes, do: "hex-bytes.too-few-bytes"
  def hex_bytes_too_many_bytes, do: "hex-bytes.too-many-bytes"
  def hex_bytes_wrong_value, do: "hex-bytes.wrong-value"

  def code_point_out_of_range, do: "code-point.out-of-range"
  def code_point_surrogate, do: "code-point.surrogate"
  def code_point_wrong_value, do: "code-point.wrong-value"

  def endianness_wrong_choice, do: "endianness.wrong-choice"

  def useful_bit_count_non_positive, do: "useful-bit-count.non-positive"
  def useful_bit_count_wrong_value, do: "useful-bit-count.wrong-value"

  def offset_out_of_range, do: "offset.out-of-range"
  def offset_wrong_value, do: "offset.wrong-value"
end
